src/test/util/setup_common.cpp

```diff
@@ -120,18 +119,39 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", "Custom data directory (default: <random_string>)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    if (!m_node.args->IsArgSet("-testdatadir")) {
+        // By default, the data directory has a random name
+        const std::string testdatadir{g_insecure_rand_ctx_temp_path.rand256().ToString().c_str()};
+        m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / testdatadir;
+    } else {
+        // Custom data directory
+        const std::string testdatadir{m_node.args->GetArg("-testdatadir").value()};
```

Comment:
Can use `GetPathArg("-testdatadir")` directly.

Also, I would check the path correctness here and error out early when it is invalid.

Reply:
Good idea, but what do you mean by path correctness?

Reply:
> Good idea, but what do you mean by path correctness?

Probably we should fail when `-testdatadir` is an empty string.